intercept/src/intercept.cpp

```diff
@@ -7244,13 +7304,212 @@ void CLIntercept::setKernelArgUSMPointer(
     }
 }
 
+void CLIntercept::dumpKernelSource(cl_kernel kernel, 
+                                   uint64_t enqueueCounter,
+                                   bool byKernelName )
+{
+    std::lock_guard<std::mutex> lock(m_Mutex);
+    std::string fileNamePrefix = "";
+    OS().GetDumpDirectoryName( sc_DumpDirectoryName, fileNamePrefix );
+    fileNamePrefix += "/Replay/Enqueue_";
+    if (byKernelName)
+        fileNamePrefix += getShortKernelName(kernel);
+    else
+        fileNamePrefix += std::to_string(enqueueCounter);
+    fileNamePrefix += "/";
+    OS().MakeDumpDirectories( fileNamePrefix );
+
+    // Get the cl_program from the cl_kernel, then extract kernel source code from the cl_program
+    CLIntercept* pIntercept = GetIntercept();
+
+    cl_program tmp_program;
+    pIntercept->dispatch().clGetKernelInfo(kernel, CL_KERNEL_PROGRAM, sizeof(cl_program), &tmp_program, nullptr);
+
+    try
+    {
+        std::string source = m_SourceStringMap.at(tmp_program);
+        std::string fileName = fileNamePrefix + "kernel.cl";
+        std::ofstream output;
+        output.open(fileName.c_str(), std::ios::out | std::ios::binary );
+        output.write(source.c_str(), source.length());
+        return;
+    } 
+    catch (std::exception const& excep)
+    {
+        (void)excep;
+    }
+
+    log("[[Warning]]: Kernel source is not available! Make sure that the kernel is compiled from source (and is not cached)\n");
+    log("Now will try to output binaries, these probably won't work on other platforms!\n");
+    
+    cl_uint num_devices;
+    clGetProgramInfo(tmp_program, CL_PROGRAM_NUM_DEVICES, sizeof(cl_uint), &num_devices, nullptr);
+
+    // Grab the device ids
+    cl_device_id* devices = new cl_device_id[num_devices];
```

Comment:
Need to free this allocation (or just use a vector instead).

Reply:
Check. In a later stage I want to remove all the manual memory allocations (new[]/delete[]). For some reason I ran into a few seg faults when I changed it before and decided to focus on the functionality first :)